### PR TITLE
Automatic readme generation with cargo-readme

### LINKED ISSUE #8: automatic readme generation with cargo-readme
We should add `cargo-readme` support for 
 - `check` as an optional argument to generate and check whether the readmes still match
 - `publish` to generate and publish the readme to crates.io

On the later we should probably think through the use cases, have it work in a few different modes: `Disabled`, `IfMissing` (only generate if not yet given), `Append` (to append to an existing `Readme.md` if found otherwise create), `ForceOverwrite` (to always overwrite). We might also want to allow for a global template to be used.

We should piggy back on the existing library if possible, not run it as a subcommand. Until the feature is stable we should feature-gate it behind a non-default-feature. Once we are happy it works as expected, we activate it.

### PR DESCRIPTION
**Pull request by stiiifff**
Fixes #8

### CODE REVIEW

**Comment by stiiifff** [comment]
@gnunicorn Here's what's implemented so far:

- **check & em-dragons commands**:  added **--check-readme** flag to check whether a package's Readme  is up-to-date, outdated or missing.
- **gen-readme command**: new command to generate Readme files based on a cratet's doc comments (using cargo-readme). The **--readme-mode** (IfMissing, Append, Overwrite) allows to alter behaviour regarding the existing Readme file (if any).
- **Template support**: the readme generation will look for a template file named **README.tpl** starting from a crate's folder, and then going up until if finds one (or none). This allows to specify a template for a set of crates in the same folder (e.g. FRAME pallets, add a template in the _frame_ folder), and also to override a more general template with a more specific one in a lower-level folder. This can be useful for altering the Readme's form / content, catering for different licenses for different parts of the system etc.
- **gen-readme feature**: the gen-readme is feature-gated. The **--check-readme** flag is not (not sure how to do that), so there is a verification done if the flag is used while the feature was not compiled-in.

File: src/commands/readme.rs [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+impl Display for CheckReadmeResult {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{}",
+            match self {
+                Self::Skipped => "Skipped",
+                Self::Missing => "Missing",
+                Self::UpdateNeeded => "Updated needed",
```
Comment (gnunicorn) [inline src/commands/readme.rs:27]: typo
```suggestion
                Self::UpdateNeeded => "Update needed",
```

File: src/commands/readme.rs [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+        (mode, existing_res) => {
+            let template_path = find_readme_template(&ws.root(), &pkg_path)?;
+            c.shell()
+                .status(
+                    "Generating",
+                    format!(
+                        "Readme for {} (template: {:?})",
+                        &pkg_manifest.name(),
+                        match &template_path {
+                            Some(p) => p.strip_prefix(&root_path).unwrap().to_str().unwrap(),
```
Comment (gnunicorn) [inline src/commands/readme.rs:115]: can we be sure this never fails? why not use `?`?
```suggestion
                            Some(p) => p.strip_prefix(&root_path)?.to_str()?,
```

File: src/commands/readme.rs [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+        return Ok(lib_rs);
+    }
+
+    // try main.rs
+    let main_rs = current_dir.join("src/main.rs");
+    if main_rs.exists() {
+        return Ok(main_rs);
+    }
+
+    // try lib defined in `Cargo.toml`
```
Comment (gnunicorn) [inline src/commands/readme.rs:183]: is this still expected to be here? Either activate or remove maybe?

**Comment by gnunicorn** [comment]
@stiiifff 

it is currently failing for me on substrate master with:

```
❯ ../cargo-unleash/target/release/cargo-unleash gen-readme --readme-mode IfMissing
   Preparing Disabling Dev Dependencies
    Patching node-template
    Patching node-template-runtime
    Patching pallet-aura
    Patching frame-support
    Patching frame-metadata
    Patching sp-core
    Patching sp-debug-derive
    Patching sp-externalities
    Patching sp-std
    Patching sp-storage
    Patching sp-runtime-interface
    Patching sp-runtime-interface-proc-macro
    Patching sp-tracing
    Patching sp-wasm-interface
    Patching sp-io
    Patching sp-state-machine
    Patching sp-panic-handler
    Patching sp-trie
    Patching sp-runtime
    Patching sp-application-crypto
    Patching sp-arithmetic
    Patching sp-inherents
    Patching substrate-wasm-builder-runner
    Patching sp-serializer
    Patching frame-support-procedural
    Patching frame-support-procedural-tools
    Patching frame-support-procedural-tools-derive
    Patching frame-system
    Patching sp-version
    Patching sc-block-builder
    Patching sc-client-api
    Patching substrate-prometheus-endpoint
    Patching sc-executor
    Patching sc-executor-common
    Patching sp-allocator
    Patching sc-executor-wasmi
    Patching sc-executor-wasmtime
    Patching sp-api
    Patching sp-api-proc-macro
    Patching sp-sandbox
    Patching frame-executive
    Patching pallet-balances
    Patching frame-benchmarking
    Patching pallet-transaction-payment
    Patching pallet-transaction-payment-rpc-runtime-api
    Patching pallet-indices
    Patching sp-keyring
    Patching frame-system-rpc-runtime-api
    Patching pallet-babe
    Patching pallet-session
    Patching pallet-timestamp
    Patching sp-timestamp
    Patching sp-session
    Patching sp-staking
    Patching sp-consensus-babe
    Patching sp-consensus
    Patching sp-utils
    Patching sp-consensus-vrf
    Patching sc-service
    Patching sc-chain-spec
    Patching sc-chain-spec-derive
    Patching sc-network
    Patching fork-tree
    Patching sc-peerset
    Patching sp-blockchain
    Patching sp-block-builder
    Patching sc-telemetry
    Patching sp-chain-spec
    Patching sc-client-db
    Patching sc-state-db
    Patching sp-database
    Patching sc-informant
    Patching sp-transaction-pool
    Patching sc-keystore
    Patching sc-light
    Patching sc-offchain
    Patching sp-offchain
    Patching sc-transaction-pool
    Patching sc-transaction-graph
    Patching sc-rpc
    Patching sc-rpc-api
    Patching sp-rpc
    Patching sc-rpc-server
    Patching sc-tracing
    Patching sc-finality-grandpa
    Patching sc-consensus
    Patching sc-network-gossip
    Patching sp-finality-grandpa
    Patching sp-finality-tracker
    Patching sp-consensus-aura
    Patching pallet-grandpa
    Patching pallet-finality-tracker
    Patching pallet-offences
    Patching pallet-staking
    Patching pallet-authorship
    Patching sp-authorship
    Patching sp-npos-elections
    Patching sp-npos-elections-compact
    Patching substrate-test-utils
    Patching pallet-staking-reward-curve
    Patching pallet-randomness-collective-flip
    Patching pallet-sudo
    Patching pallet-template
    Patching sc-basic-authorship
    Patching sc-proposer-metrics
    Patching sc-cli
    Patching sc-consensus-aura
    Patching sc-consensus-slots
    Patching substrate-build-script-utils
    Patching node-bench
    Patching node-primitives
    Patching node-runtime
    Patching frame-system-benchmarking
    Patching pallet-authority-discovery
    Patching sp-authority-discovery
    Patching pallet-collective
    Patching pallet-contracts
    Patching pallet-contracts-primitives
    Patching pallet-contracts-rpc-runtime-api
    Patching pallet-democracy
    Patching pallet-scheduler
    Patching pallet-elections-phragmen
    Patching pallet-identity
    Patching pallet-im-online
    Patching pallet-membership
    Patching pallet-multisig
    Patching pallet-offences-benchmarking
    Patching pallet-proxy
    Patching pallet-utility
    Patching pallet-recovery
    Patching pallet-session-benchmarking
    Patching pallet-society
    Patching pallet-treasury
    Patching pallet-vesting
    Patching node-testing
    Patching node-executor
    Patching node-browser-testing
    Patching node-cli
    Patching substrate-browser-utils
    Patching frame-benchmarking-cli
    Patching node-inspect
    Patching node-rpc
    Patching pallet-contracts-rpc
    Patching pallet-transaction-payment-rpc
    Patching sc-consensus-babe
    Patching sc-consensus-epochs
    Patching sc-consensus-uncles
    Patching sc-consensus-babe-rpc
    Patching sc-finality-grandpa-rpc
    Patching substrate-frame-rpc-system
    Patching sc-authority-discovery
    Patching node-rpc-client
    Patching subkey
    Patching chain-spec-builder
    Patching sc-consensus-manual-seal
    Patching sc-consensus-pow
    Patching sp-consensus-pow
    Patching pallet-assets
    Patching pallet-benchmark
    Patching pallet-elections
    Patching pallet-evm
    Patching pallet-example
    Patching pallet-example-offchain-worker
    Patching pallet-generic-asset
    Patching pallet-nicks
    Patching pallet-scored-pool
    Patching substrate-frame-rpc-support
    Patching substrate-wasm-builder
   Resolving Dependency Tree
     Syncing Versions from crates.io
  Generating Readme files
  Generating Readme for pallet-template (template: "none found")
  Generating Readme for node-template-runtime (template: "none found")
  Generating Readme for node-template (template: "none found")
Error: "Failure generating Readme for node-template: missing field `path` for key `bin` at line 44 column 1"
```

**Comment by stiiifff** [comment]
@gnunicorn Let me check why it's failing on master.

File: src/commands/readme.rs (hunk 3 above)
Comment (stiiifff) [reply src/commands/readme.rs]: See below.

**Comment by stiiifff** [comment]
Seems related to https://github.com/livioribeiro/cargo-readme/pull/17, will fix it locally and push them upstream later.

**Comment by stiiifff** [comment]
Actually, this will need to be fixed in **cargo-readme**, no easy local fix. I'll look into it.

Meanwhile, I suggest **skipping those crates that have a `[lib]`** table in their manifest:
`
../cargo-unleash/target/release/cargo-unleash gen-readme --skip 'node-*' sp-debug-derive sp-runtime-interface-proc-macro sp-core frame-support-procedural-tools-derive frame-support-procedural sp-api-proc-macro sc-chain-spec-derive sp-npos-elections-compact pallet-staking-reward-curve --readme-mode IfMissing
`

Also, as I am using `commands::packages_to_release` to get the list of packages, you **first need to bump the version** before (re-)generating the READMEs e.g. `../cargo-unleash/target/release/cargo-unleash version set-pre "2.0.0-rc4"`.

With that, I can generate the readmes on master, let me know how it goes on your end.

**Pull request merged**